File: go/doubly_linked_list.go
package main

import (
	"fmt"
)

type Node struct {
	key int
	prev *Node
	next *Node
}

type DoublyLinkedList struct {
	head *Node
}

func (doublyLinkedList *DoublyLinkedList) push(value int) {
	newNode := &Node{
		key: value,
	}
	newNode.next = doublyLinkedList.head

	if doublyLinkedList.head != nil {
		doublyLinkedList.head.prev = newNode
	}

	doublyLinkedList.head = newNode
}

func (doublyLinkedList *DoublyLinkedList) insertAfter(prevNode *Node, value int) {
	newNode := &Node {
		key: value,
	}
	newNode.next = prevNode.next
	prevNode.next = newNode
	newNode.prev = prevNode

	if newNode.next != nil {
		newNode.next.prev = newNode
	}
}

func (doublyLinkedList *DoublyLinkedList) append(value int) {

	newNode := &Node {
		key: value,
	}
	curr := doublyLinkedList.head

	if curr == nil {
		doublyLinkedList.head = newNode
		return
	}

	for curr.next != nil {
		curr = curr.next
	}

	curr.next = newNode
	newNode.prev = curr

}

func (doublyLinkedList *DoublyLinkedList) print() {
	curr := doublyLinkedList.head

	for curr != nil {
		fmt.Println(curr.key)
		curr = curr.next
	}
	fmt.Println("-")
}

func (doublyLinkedList *DoublyLinkedList) delete(value int) {
	curr := doublyLinkedList.head

	if curr != nil && curr.key == value {
		doublyLinkedList.head = curr.next
		if doublyLinkedList.head != nil {
			doublyLinkedList.head.prev = nil
		}
		return
	}

	for curr != nil {
		if curr.key == value {
			curr.prev.next = curr.next
			if curr.next != nil {
				curr.next.prev = curr.prev
			}
		}
		curr = curr.next
	}

}

func main() {

	doublyLinkedList := DoublyLinkedList{}

	doublyLinkedList.push(2)
	doublyLinkedList.print()
	doublyLinkedList.push(7)
	doublyLinkedList.print()
	doublyLinkedList.push(8)
	doublyLinkedList.print()
	doublyLinkedList.push(10)
	doublyLinkedList.print()
	doublyLinkedList.append(4)
	doublyLinkedList.print()
	doublyLinkedList.delete(8)
	doublyLinkedList.print()
}

File: python/doubly_linked_list.py
class Node:

    def __init__(self, prev=None, next=None, data=None):
        self.data = data
        self.prev = prev
        self.next = next


class DoublyLinkedList:

    def __init__(self):
        self.head = None

    def push(self, new_data):
        new_node = Node(data=new_data)
        new_node.next = self.head

        if self.head:
            self.head.prev = new_node

        self.head = new_node

    def insert_after(self, new_data, prev_node):
        new_node = Node(data=new_data)
        new_node.next = prev_node.next
        prev_node.next = new_node
        new_node.prev = prev_node

        if new_node.next:
            new_node.next.prev = new_node

    def append(self, new_data):
        new_node = Node(data=new_data)
        if not self.head:
            self.head = new_node
            return

        last = self.head
        while(last.next):
            last = last.next

        last.next = new_node
        new_node.prev = last

    def print_list(self):
        temp = self.head
        while(temp):
            print(temp.data)
            temp = temp.next

    def delete_node(self, key):
        temp = self.head

        if temp and temp.data == key:
            self.head = temp.next
            if self.head:
                self.head.prev = None
            temp = None
            return

        while(temp):
            if temp.data == key:
                break
            temp = temp.next

        if not temp:
            return

        temp.prev.next = temp.next
        if temp.next:
            temp.next.prev = temp.prev
        temp = None


if __name__ == '__main__':

    dll = DoublyLinkedList()
    dll.push(2)
    dll.push(7)
    dll.push(8)
    dll.push(10)
    dll.append(4)
    dll.insert_after(3, dll.head.next)
    dll.print_list()
    dll.delete_node(10)
    print('-')
    dll.print_list()
